File: android/app/src/main/kotlin/com/privateagent/private_agent/service/RecoveryEngine.kt
package com.privateagent.private_agent.service

import com.privateagent.private_agent.model.ScreenNode
import com.privateagent.private_agent.model.TaskStep

class RecoveryEngine(
    private val maxRepeatingSteps: Int = 3,
    private val maxUnchangedScreens: Int = 3
) {
    private val actionHistory = mutableListOf<String>()
    private val screenSignatures = mutableListOf<String>()
    private var recoveryAttempt = 0

    fun reset() {
        actionHistory.clear()
        screenSignatures.clear()
        recoveryAttempt = 0
    }

    fun recordAction(step: TaskStep) {
        actionHistory.add(step.signature)
        if (actionHistory.size > 10) {
            actionHistory.removeAt(0)
        }
    }

    fun recordScreen(nodes: List<ScreenNode>) {
        val sig = nodes.joinToString("|") { "${it.index}:${it.bestLabel}:${it.centerX},${it.centerY}" }
        screenSignatures.add(sig)
        if (screenSignatures.size > 10) {
            screenSignatures.removeAt(0)
        }
    }

    fun isStuck(): Boolean {
        // Check repeating actions
        if (actionHistory.size >= maxRepeatingSteps) {
            val tail = actionHistory.takeLast(maxRepeatingSteps)
            if (tail.distinct().size == 1) {
                return true
            }
        }

        // Check unchanged screens
        if (screenSignatures.size >= maxUnchangedScreens) {
            val tail = screenSignatures.takeLast(maxUnchangedScreens)
            if (tail.distinct().size == 1 && tail.first().isNotBlank()) {
                return true
            }
        }

        return false
    }

    fun getSuggestedRecoveryStep(): TaskStep {
        recoveryAttempt++
        return when (recoveryAttempt % 3) {
            1 -> TaskStep(
                action = TaskStep.ACTION_SCROLL,
                params = mapOf("direction" to "down"),
                reasoning = "Recovery fallback: Screen appears stuck or repeating actions. Scrolling down to reveal more options."
            )
            2 -> TaskStep(
                action = TaskStep.ACTION_WAIT,
                params = mapOf("seconds" to 2),
                reasoning = "Recovery fallback: Waiting for UI animations or async loading."
            )
            else -> TaskStep(
                action = TaskStep.ACTION_PRESS_BACK,
                params = emptyMap(),
                reasoning = "Recovery fallback: Dismissing potential dialog or keyboard blocking the target."
            )
        }
    }
}

File: lib/services/recovery_engine.dart
import '../models/task_step.dart';

/// Detects when the agent loop is stuck (same action repeating, or the
/// screen never changing) and proposes corrective fallback actions.
class RecoveryEngine {
  static const _signatureWindow = 3;
  static const _unchangedThreshold = 3;

  final List<String> _recentSignatures = [];
  String? _lastScreenFingerprint;
  int _unchangedScreens = 0;
  int _recoveryCursor = 0;

  void reset() {
    _recentSignatures.clear();
    _lastScreenFingerprint = null;
    _unchangedScreens = 0;
    _recoveryCursor = 0;
  }

  void recordAction(TaskStep step) {
    _recentSignatures.add(step.signature);
    if (_recentSignatures.length > 10) {
      _recentSignatures.removeAt(0);
    }
  }

  /// Records a lightweight fingerprint of the latest screen dump.
  void recordScreen(String formattedDump) {
    if (formattedDump == _lastScreenFingerprint) {
      _unchangedScreens++;
    } else {
      _unchangedScreens = 0;
      _lastScreenFingerprint = formattedDump;
    }
  }

  /// The exact same action (name + params) repeated >= 3 times in a row.
  bool get isRepeatingActions {
    if (_recentSignatures.length < _signatureWindow) return false;
    final tail =
        _recentSignatures.sublist(_recentSignatures.length - _signatureWindow);
    return tail.toSet().length == 1;
  }

  bool get isScreenStuck => _unchangedScreens >= _unchangedThreshold;

  bool get shouldRecover => isRepeatingActions || isScreenStuck;

  /// Rotating recovery strategies: scroll for more content, wait for slow
  /// screens, then back out of a dead end.
  TaskStep suggestRecovery() {
    const strategies = [
      TaskStep(
        action: TaskStep.actionScroll,
        params: {'direction': 'down'},
        reasoning: 'Recovery: screen appears stuck; scrolling for new content.',
      ),
      TaskStep(
        action: TaskStep.actionWait,
        params: {'ms': 2000},
        reasoning: 'Recovery: waiting for the UI to settle.',
      ),
      TaskStep(
        action: TaskStep.actionPressBack,
        params: {},
        reasoning: 'Recovery: backing out of a possible dead end.',
      ),
    ];
    final step = strategies[_recoveryCursor % strategies.length];
    _recoveryCursor++;
    return step;
  }
}
